### ISSUE TITLE
Linux version

### ISSUE THREAD

**Issue by vzaliva** [ahrm/sioyek#1]
Please make it available on Linux! Thanks!

**Comment by mnafees** [comment]
Would love to help you port it to Linux as I see that the project build files seem to be generated with Qt qmake!

**Comment by ahrm** [comment]
> Would love to help you port it to Linux as I see that the project build files seem to be generated with Qt qmake!

The project currently builds and works successfully on linux. What I am stuck with is packaging the executable along with the dependencies so the user doesn't have to manually install libraries.

It probably is embarrassingly easy to do but I am a noob when it comes to linux builds.

**Comment by Ristovski** [comment]
What I would suggest is creating a static build with Github Actions or packagaing an AppImage.
I would take a look at these, but not right now as I am sadly too busy.

**Comment by mnafees** [comment]
> > Would love to help you port it to Linux as I see that the project build files seem to be generated with Qt qmake!
> 
> 
> 
> The project currently builds and works successfully on linux. What I am stuck with is packaging the executable along with the dependencies so the user doesn't have to manually install libraries.
> 
> 
> 
> It probably is embarrassingly easy to do but I am a noob when it comes to linux builds.

I believe that for Qt apps on Linux that target an AppImage, https://github.com/probonopd/linuxdeployqt has been the go to. Maybe you can check it out? It's also linked to in https://wiki.qt.io/Deploying_a_Qt5_Application_Linux

**Comment by vzaliva** [comment]
Many users will prefer APT package for Ubuntu/Debian and other distributions using APT packages. This will also take care of dependencies.

**Comment by khughitt** [comment]
Has anyone had any luck building this for Arch by chance? After installing qt5-3d, I was able to generate a Makefile, but run into linker errors when attempting to run make:

```
cd pdf_viewer
qmake -o Makefile pdf_viewer.pro
make
```

Output for make:

```
g++ -Wl,-rpath-link,/usr/lib -o ../x64/Debug/pdf_viewer    /usr/lib/libQt5OpenGL.so/usr/lib/libQt5QuickWidgets.so /usr/lib/libQt5Widgets.so /usr/lib/libQt5Quick.so /usr/lib/libQt53DAnimation.so /usr/lib/libQt53DExtras.so /usr/lib/libQt53DRender.so /usr/lib/libQt53DInput.so /usr/lib/libQt53DLogic.so /usr/lib/libQt53DCore.so /usr/lib/libQt5OpenGLExtensions.a /usr/lib/libQt5Gui.so /usr/lib/libQt5Sql.so /usr/lib/libQt5QmlModels.so /usr/lib/libQt5Qml.so /usr/lib/libQt5Network.so /usr/lib/libQt5Core.so -lGL -lpthread
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/Scrt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
collect2: error: ld returned 1 exit status
make: *** [Makefile:282: ../x64/Debug/pdf_viewer] Error 1
```

**Comment by shenlebantongying** [comment]
There is a similar program that can run on Linux
https://en.wikipedia.org/wiki/Zathura_(document_viewer)
https://pwmt.org/projects/zathura/screenshots/

**Comment by ahrm** [comment]
> Has anyone had any luck building this for Arch by chance? After installing qt5-3d, I was able to generate a Makefile, but run into linker errors when attempting to run make:
> 
> ```
> cd pdf_viewer
> qmake -o Makefile pdf_viewer.pro
> make
> ```
> 
> Output for make:
> 
> ```
> g++ -Wl,-rpath-link,/usr/lib -o ../x64/Debug/pdf_viewer    /usr/lib/libQt5OpenGL.so/usr/lib/libQt5QuickWidgets.so /usr/lib/libQt5Widgets.so /usr/lib/libQt5Quick.so /usr/lib/libQt53DAnimation.so /usr/lib/libQt53DExtras.so /usr/lib/libQt53DRender.so /usr/lib/libQt53DInput.so /usr/lib/libQt53DLogic.so /usr/lib/libQt53DCore.so /usr/lib/libQt5OpenGLExtensions.a /usr/lib/libQt5Gui.so /usr/lib/libQt5Sql.so /usr/lib/libQt5QmlModels.so /usr/lib/libQt5Qml.so /usr/lib/libQt5Network.so /usr/lib/libQt5Core.so -lGL -lpthread
> /usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/Scrt1.o: in function `_start':
> (.text+0x24): undefined reference to `main'
> collect2: error: ld returned 1 exit status
> make: *** [Makefile:282: ../x64/Debug/pdf_viewer] Error 1
> ```

For linux builds I use the attached .pro file (the .pro file in the main repo is generated by visual studio). You need to manually install all the required libraries.

(I had to change the file extension to .txt so I could attach it in this issue)
[pdf_viewer.txt](https://github.com/ahrm/sioyek/files/6850570/pdf_viewer.txt)

**Comment by ahrm** [comment]
> Has anyone had any luck building this for Arch by chance? After installing qt5-3d, I was able to generate a Makefile, but run into linker errors when attempting to run make:
> 
> ```
> cd pdf_viewer
> qmake -o Makefile pdf_viewer.pro
> make
> ```
> 
> Output for make:
> 
> ```
> g++ -Wl,-rpath-link,/usr/lib -o ../x64/Debug/pdf_viewer    /usr/lib/libQt5OpenGL.so/usr/lib/libQt5QuickWidgets.so /usr/lib/libQt5Widgets.so /usr/lib/libQt5Quick.so /usr/lib/libQt53DAnimation.so /usr/lib/libQt53DExtras.so /usr/lib/libQt53DRender.so /usr/lib/libQt53DInput.so /usr/lib/libQt53DLogic.so /usr/lib/libQt53DCore.so /usr/lib/libQt5OpenGLExtensions.a /usr/lib/libQt5Gui.so /usr/lib/libQt5Sql.so /usr/lib/libQt5QmlModels.so /usr/lib/libQt5Qml.so /usr/lib/libQt5Network.so /usr/lib/libQt5Core.so -lGL -lpthread
> /usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/Scrt1.o: in function `_start':
> (.text+0x24): undefined reference to `main'
> collect2: error: ld returned 1 exit status
> make: *** [Makefile:282: ../x64/Debug/pdf_viewer] Error 1
> ```

I added the linux .pro file to the main repo in e4656e7154d48b522d7c722c3eebcdfab47e5d00. Also there was some compile errors with g++ not implicitly converting `std::filesystem::path` to `std::wstring` which is fixed in this commit.

**Comment by vestingz** [comment]
Thanks for working to add support for Linux. Compiling with gcc v11.1.0 and libmupdf 1.18.0 returns following error:

```
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_lcms_log_error':
(.text.fz_lcms_log_error+0x7): undefined reference to `cmsGetContextUserData'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_lcms_realloc':
(.text.fz_lcms_realloc+0xd): undefined reference to `cmsGetContextUserData'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_lcms_free':
(.text.fz_lcms_free+0x6): undefined reference to `cmsGetContextUserData'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_lcms_malloc':
(.text.fz_lcms_malloc+0x5): undefined reference to `cmsGetContextUserData'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_drop_icc_link_imp':
(.text.fz_drop_icc_link_imp+0x1e): undefined reference to `cmsDeleteTransform'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_new_icc_context':
(.text.fz_new_icc_context+0x10): undefined reference to `cmsCreateContext'
/usr/bin/ld: (.text.fz_new_icc_context+0x31): undefined reference to `cmsSetLogErrorHandler'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_drop_icc_context':
(.text.fz_drop_icc_context+0x16): undefined reference to `cmsDeleteContext'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_new_icc_profile':
(.text.fz_new_icc_profile+0x11): undefined reference to `cmsOpenProfileFromMem'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_icc_profile_is_lab':
(.text.fz_icc_profile_is_lab+0x16): undefined reference to `cmsGetColorSpace'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_drop_icc_profile':
(.text.fz_drop_icc_profile+0x12): undefined reference to `cmsCloseProfile'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_icc_profile_name':
(.text.fz_icc_profile_name+0x1f): undefined reference to `cmsReadTag'
/usr/bin/ld: (.text.fz_icc_profile_name+0x46): undefined reference to `cmsMLUgetASCII'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_icc_profile_components':
(.text.fz_icc_profile_components+0x11): undefined reference to `cmsGetColorSpace'
/usr/bin/ld: (.text.fz_icc_profile_components+0x1d): undefined reference to `cmsChannelsOf'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_new_icc_link':
(.text.fz_new_icc_link+0x97): undefined reference to `cmsGetColorSpace'
/usr/bin/ld: (.text.fz_new_icc_link+0xaa): undefined reference to `_cmsLCMScolorSpace'
/usr/bin/ld: (.text.fz_new_icc_link+0xbd): undefined reference to `cmsChannelsOf'
/usr/bin/ld: (.text.fz_new_icc_link+0x10e): undefined reference to `cmsGetColorSpace'
/usr/bin/ld: (.text.fz_new_icc_link+0x120): undefined reference to `_cmsLCMScolorSpace'
/usr/bin/ld: (.text.fz_new_icc_link+0x135): undefined reference to `cmsChannelsOf'
/usr/bin/ld: (.text.fz_new_icc_link+0x1db): undefined reference to `cmsGetColorSpace'
/usr/bin/ld: (.text.fz_new_icc_link+0x1ee): undefined reference to `_cmsLCMScolorSpace'
/usr/bin/ld: (.text.fz_new_icc_link+0x201): undefined reference to `cmsChannelsOf'
/usr/bin/ld: (.text.fz_new_icc_link+0x239): undefined reference to `cmsCreateTransform'
/usr/bin/ld: (.text.fz_new_icc_link+0x25e): undefined reference to `cmsTransform2DeviceLink'
/usr/bin/ld: (.text.fz_new_icc_link+0x270): undefined reference to `cmsDeleteTransform'
/usr/bin/ld: (.text.fz_new_icc_link+0x2c1): undefined reference to `cmsCreateMultiprofileTransform'
/usr/bin/ld: (.text.fz_new_icc_link+0x2dc): undefined reference to `cmsCloseProfile'
/usr/bin/ld: (.text.fz_new_icc_link+0x371): undefined reference to `cmsCreateTransform'
/usr/bin/ld: (.text.fz_new_icc_link+0x416): undefined reference to `cmsCreateTransform'
/usr/bin/ld: (.text.fz_new_icc_link+0x46e): undefined reference to `cmsCreateTransform'
/usr/bin/ld: (.text.fz_new_icc_link+0x4a6): undefined reference to `cmsDeleteTransform'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_icc_transform_color':
(.text.fz_icc_transform_color+0x96): undefined reference to `cmsDoTransform'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_icc_transform_pixmap':
(.text.fz_icc_transform_pixmap+0xaa): undefined reference to `cmsGetTransformInputFormat'
/usr/bin/ld: (.text.fz_icc_transform_pixmap+0xc1): undefined reference to `cmsGetTransformOutputFormat'
/usr/bin/ld: (.text.fz_icc_transform_pixmap+0x172): undefined reference to `cmsDoTransform'
/usr/bin/ld: (.text.fz_icc_transform_pixmap+0x2cf): undefined reference to `cmsDoTransform'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(xml.o): in function `xml_from_gumbo':
(.text.xml_from_gumbo+0xc2): undefined reference to `gumbo_normalized_tagname'
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(xml.o): in function `fz_parse_xml_from_html5':
(.text.fz_parse_xml_from_html5+0x124): undefined reference to `gumbo_destroy_output'
/usr/bin/ld: (.text.fz_parse_xml_from_html5+0x23a): undefined reference to `gumbo_parse_with_options'
collect2: Fehler: ld gab 1 als Ende-Status zurück
make: *** [Makefile:284: pdf_viewer_project] Fehler 1
```

**Comment by khughitt** [comment]
Similar results here. lcms/gumbo versions on system:

- libgumbo.so.1.0.0 (0.10.1)
- liblcms2.so.2.0.12 (2.12)

**Comment by ahrm** [comment]
I made an experimental linux build which is available here:

https://github.com/ahrm/sioyek/releases/tag/v0.31.3

Hopefully it works on your systems too. Here is my build method and software versions that I used for this build:

- Ubuntu 20.10 groovy
- Qt 5.14.2
- libmupdf 1.17.0

Build instructions:
* Download the requires libraries:
`sudo apt install libmupdf-dev libzip-dev libfreetype-dev libmujs-dev libgif-dev libjbig2dec0-dev libopenjp2-7-dev libjpeg-dev libhardfbuzz-dev`
* Clone the repository
```
git clone https://github.com/ahrm/sioyek
cd sioyek/pdf_viewer
```
* Create makefiles and run makefiles:
```
qmake pdf_viewer_linux.pro
make
```

I was not able to reproduce your errors on my machine, I will try to make an arch virtual machine to see if I can reproduce the errors.

**Comment by vestingz** [comment]
Thanks for providing the build! Unfortunately, it doesn't really work on my machine, so I opened a new issue #4 .

**Comment by ahrm** [comment]
> Thanks for providing the build! Unfortunately, it doesn't really work on my machine, so I opened a new issue #4 .

Thanks for your feedback. I am making an arch virtual machine to investigate the issue.

**Comment by vzaliva** [comment]
```
$sudo apt install libmupdf-dev libzip-dev libfreetype-dev libmujs-dev libgif-dev libjbig2dec0-dev libopenjp2-7-dev libjpeg-dev libhardfbuzz-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package libmujs-dev
E: Unable to locate package libhardfbuzz-dev
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal

```

**Comment by ahrm** [comment]
> Unable to locate package libmujs-dev

Sorry, I had a typo it should be `libharfbuzz-dev` not `libhardfbuzz-dev`. Apparently libmujs-dev is not available on ubuntu 20.04. I will investigate solutions.

**Comment by ahrm** [comment]
> Thanks for working to add support for Linux. Compiling with gcc v11.1.0 and libmupdf 1.18.0 returns following error:
> 
> ```
> /usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_lcms_log_error':
> (.text.fz_lcms_log_error+0x7): undefined reference to `cmsGetContextUserData'
> /usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(color-lcms.o): in function `fz_lcms_realloc':
> (.text.fz_lcms_realloc+0xd): undefined reference to `cmsGetContextUserData'
...
...
...
> /usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../lib/libmupdf.a(xml.o): in function `fz_parse_xml_from_html5':
> (.text.fz_parse_xml_from_html5+0x124): undefined reference to `gumbo_destroy_output'
> /usr/bin/ld: (.text.fz_parse_xml_from_html5+0x23a): undefined reference to `gumbo_parse_with_options'
> collect2: Fehler: ld gab 1 als Ende-Status zurück
> make: *** [Makefile:284: pdf_viewer_project] Fehler 1
> ```

So apparently mupdf 1.18.0 has two new dependencies: lcms2 and gumbo-parser. I was able to fix the issue by first installing these:
`pacman -S lcms2 gumbo-parser`
and then updating the linked libraries in `pdf_viewer_linux.pro`
from 
`LIBS += -ldl -lmupdf -lz -lfreetype -lmujs -lgif -ljbig2dec -lopenjp2 -ljpeg -lharfbuzz`
to
`LIBS += -ldl -lmupdf -lz -lfreetype -lmujs -lgif -ljbig2dec -lopenjp2 -ljpeg -lharfbuzz -llcms2 -lgumbo`

**Comment by Omar-Elrefaei** [comment]
> (BTW please tell me if the experimental build works on your machines or not)

The build workes perfectly on my wonky (nowhere near default, many libs installed) Ubuntu 21.04 machine 👍🏼

**Comment by s1291** [comment]
On Ubuntu 20.04, using Qt 5.11, I can't get it to compile. Following the steps above, I get the following error:
 ```
qmake pdf_viewer.pro
Project ERROR: Unknown module(s) in QT: 3dcore 3danimation 3dextras 3dinput 3dlogic 3drender
```

**Comment by ahrm** [comment]
> On Ubuntu 20.04, using Qt 5.12.8, I can't get it to compile. Following the steps above, I get the following error:
> 
> ```
> make
> 
> ....
> ....
> pdf_viewer/document.cpp: In member function ‘void Document::count_chapter_pages(std::vector<int, std::allocator<int> >&)’:
> pdf_viewer/document.cpp:146:21: error: ‘fz_count_chapters’ was not declared in this scope; did you mean ‘fz_count_pages’?
>  146 |  int num_chapters = fz_count_chapters(context, doc);
>      |                     ^~~~~~~~~~~~~~~~~
>      |                     fz_count_pages
> pdf_viewer/document.cpp:149:19: error: ‘fz_count_chapter_pages’ was not declared in this scope; did you mean ‘count_chapter_pages’?
>  149 |   int num_pages = fz_count_chapter_pages(context, doc, i);
>      |                   ^~~~~~~~~~~~~~~~~~~~~~
>      |                   count_chapter_pages
> pdf_viewer/document.cpp: In member function ‘void Document::count_chapter_pages_accum(std::vector<int, std::allocator<int> >&)’:
> pdf_viewer/document.cpp:161:20: warning: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<int, std::allocator<int> >::size_type’ {aka ‘long unsigned int’} [-Wsign-compare]
>  161 |  for (int i = 0; i < raw_page_count.size(); i++) {
>      |                  ~~^~~~~~~~~~~~~~~~~~~~~~~
> pdf_viewer/document.cpp: In member function ‘void Document::convert_toc_tree(fz_outline*, std::vector<TocNode*>&)’:
> pdf_viewer/document.cpp:316:4: error: ‘fz_location’ was not declared in this scope
>  316 |    fz_location loc = fz_resolve_link(context, doc, root->uri, &xp, &yp);
>      |    ^~~~~~~~~~~
> pdf_viewer/document.cpp:317:43: error: ‘loc’ was not declared in this scope; did you mean ‘doc’?
>  317 |    int chapter_page = accum_chapter_pages[loc.chapter];
> 
> ...
> ```

You are using an old version of libmupdf (probably older than 1.17.0). I have included mupdf as a submodule to fix these problems with different mupdf versions in different operating systems. Try the following procedure:

First, install libharfbuzz:
`sudo apt install libharfbuzz-dev`

Now clone the repository:
`git clone --recursive https://github.com/ahrm/sioyek`

And run the helper build script:
```
cd sioyek
./build_linux.sh
```

**Comment by vzaliva** [comment]
Can somebody produce ubuntu package to take care of all dependencies and build complications? It would simplify things greatly. I am personally interested in Ubuntu support, but the same package could be then easily adopted to many apt-based distros.

**Comment by ahrm** [comment]
> Can somebody produce ubuntu package to take care of all dependencies and build complications? It would simplify things greatly. I am personally interested in Ubuntu support, but the same package could be then easily adopted to many apt-based distros.

I have a test release package here:

https://workupload.com/file/3tNQxD5zQuZ

does this work for you?

**Comment by vzaliva** [comment]
not sure how it install and run it. If you expect users to untar it from / this is a no-go. Installing shared libraries in /usr/lib could mess up your system and presents a security concern. Also, it raises the question of how to uninstall. 100% software I use and I think am not unique here is installed in one of 3 ways:

1. via `apt` (preferred way)
2. via `snap`
3. compiled manually with `configure --root=/usr/local` (and installed via `make install`). (least preferable)

**Comment by ahrm** [comment]
> not sure how it install and run it. If you expect users to untar it from / this is a no-go. Installing shared libraries in /usr/lib could mess up your system and presents a security concern. Also, it raises the question of how to uninstall. 100% software I use and I think am not unique here is installed in one of 3 ways:
> 
> 1. via `apt` (preferred way)
> 2. via `snap`
> 3. compiled manually with `configure --root=/usr/local` (and installed via `make install`). (least preferable)

You are right, as I said, the linux build currently is experimental so there are a few inconveniences.
With this package you don't have to install dependencies though. Just unzip and run `./AppRun`.

**Comment by vzaliva** [comment]
It briefly shows an empty window then crash with:

fish: Job 4, './AppRun' terminated by signal SIGSEGV (Address boundary error)

it could be due to a mismatch with some other shared libraries. You can probably try to linkx it statically,
but frankly, the Linux way is to use a package manager which will enforce the dependencies.

**Comment by s1291** [comment]
> > Can somebody produce ubuntu package to take care of all dependencies and build complications? It would simplify things greatly. I am personally interested in Ubuntu support, but the same package could be then easily adopted to many apt-based distros.
> 
> I have a test release package here:
> 
> https://workupload.com/file/3tNQxD5zQuZ
> 
> does this work for you?

I have downloaded that but it doesn't work when I invoke `./AppRun` it shows an empty black window. The workaround is to start sioyek from `/ioyek-release/usr/bin` This will automatically opens the PDF tutorial. I have tested some features, most of them work perfectly except that:

* The G shortcut (or Shift + g ) doesn't work.
* When scrolling through pages, there is a checkerboard artifact that appears and disappears immediately.

I have also tried to compile sioyek (cloned recursively with newer version mupdf). Unfortunately, right now I am using an unstable connection and since git doesn't support resuming I can't get the repository to be cloned properly (until I have access to a decent one).

Thank you

**Comment by ahrm** [comment]
> 
> 
> > > Can somebody produce ubuntu package to take care of all dependencies and build complications? It would simplify things greatly. I am personally interested in Ubuntu support, but the same package could be then easily adopted to many apt-based distros.
> > 
> > 
> > I have a test release package here:
> > https://workupload.com/file/3tNQxD5zQuZ
> > does this work for you?
> 
> I have downloaded that but it doesn't work when I invoke `./AppRun` it shows an empty black window. The workaround is to start sioyek from `./sioyek-release/usr/bin` This will automatically open the PDF tutorial. I have tested some features, most of them work perfectly except that:
> 
>     * The G shortcut (or Shift + g ) doesn't work.
> 
>     * When scrolling through pages, there is a checkerboard artifact that appears and disappears immediately.
> 
> 
> I have also tried to compile sioyek (cloned recursively with newer version mupdf). Unfortunately, right now I am using an unstable connection, and since git doesn't support resuming I can't get the repository to be cloned properly (until I have access to a decent one).
> 
> Thank you

> I have downloaded that but it doesn't work when I invoke `./AppRun` it shows an empty black window. The workaround is to start sioyek from `./sioyek-release/usr/bin` This will automatically open the PDF tutorial.

Yes. Some other users are reporting the same experience with AppRun. I am investigating the problem.

>     * The G shortcut (or Shift + g ) doesn't work.

Oops. It appears that I have accidentally deleted the shortcut from `keys.config`. You can fix it by adding the following line to `keys.config`:
`goto_end            <S-g>`

>     * When scrolling through pages, there is a checkerboard artifact that appears and disappears immediately.
This is not a bug. The checkerboard pattern indicates that the page is not yet rendered. It is by design (unless there is some weird flickering or something going on? It should appear once when you first visit a page and then should disappear).

**Comment by khughitt** [comment]
Just a quick update: I tried running the new linux build script in 7c26891623ddc5c53d231ee3194e3c2fad0fe5a4, and get a new linking error related to mupdf:

```
/usr/bin/ld: cannot find -lmupdf-threads
collect2: error: ld returned 1 exit status
```

libmupdf: 1.18.0-2.

**Comment by ahrm** [comment]
> Just a quick update: I tried running the new linux build script in [7c26891](https://github.com/ahrm/sioyek/commit/7c26891623ddc5c53d231ee3194e3c2fad0fe5a4), and get a new linking error related to mupdf:
> 
> ```
> /usr/bin/ld: cannot find -lmupdf-threads
> collect2: error: ld returned 1 exit status
> ```
> 
> libmupdf: 1.18.0-2.

Did you build the mupdf included in the repo or used your system's mupdf? Did you clone recursively? The correct way to build is:

```
git clone --recursive https://github.com/ahrm/sioyek
cd sioyek
./build_linux.sh
```

**Comment by khughitt** [comment]
Ah nice! I completely missed that change. The build goes well after pulling the submodule.

I tried opening a couple of pdfs though, and it both cases sioyek displays a white screen with the page numbers shown, but nothing else. Initially, it does show the status as "indexing.."

Strangely, none of the keyboard shortcuts appear to work.

In the terminal, I get:

```
could not add font!1config file change listener constructor called1starting index thread ...
     2 Impossible to open Impossible to open Impossible to open Impossible to open Pg. Are you in the right directory ? Don't forget to read the FAQ !
Figure 1.1....
texture..
...

For reference, my OpenGL version is 4.6.0. 

Let me know if there is anything else I can do to help provide more information.

**Comment by ahrm** [comment]
> Ah nice! I completely missed that change. The build goes well after pulling the submodule.
> 
> I tried opening a couple of pdfs though, and it both cases sioyek displays a white screen with the page numbers shown, but nothing else. Initially, it does show the status as "indexing.."
> 
> Strangely, none of the keyboard shortcuts appear to work.
> 
> In the terminal, I get:
> 
> ```
> could not add font!1config file change listener constructor called1starting index thread ...
>      2 Impossible to open Impossible to open Impossible to open Impossible to open Pg. Are you in the right directory ? Don't forget to read the FAQ !
> Figure 1.1....
> texture..
> ...
> 
> For reference, my OpenGL version is 4.6.0. 
> 
> Let me know if there is anything else I can do to help provide more information.
> ```

There is currently a bug in the linux version which causes the tutorial document to be open in a very zoomed out state. I suspect the white screen you are seeing is a very zoomed out tutorial document (to fix it press f10 to readjust the document to window). The shortcuts should be working though, I think the shortcut issue and the impossible to open error is probably caused by launching the wrong `sioyek` executable. You should be launching the executable in the `build` folder. That is:
```
./build_linux.sh
./build/sioyek
```

**Comment by khughitt** [comment]
Okay great! It appears to be working well for me now! The only things I've noticed so far, which is mostly just an aesthetic issue, is the appearance of a black/white checkerboarding between page renders, especially if I am jumping around fast.

All-in-all, seems to work quite well though! Thanks for taking the time to help get the linux builds working :)

**Comment by ahrm** [comment]
I have added a new binary release (AppImage) for linux here:

https://github.com/ahrm/sioyek/releases/tag/v0.31.4

I think I have fixed most bugs and compatibility issues, so it should work fine on most systems. I will keep this issue open for a couple of days in case there is any unexpected bugs but otherwise this issue is closed as far as I am concerned.

Unfortunately I do not have the time and energy to do a separate packages for debian, redhat, etc. but if anyone is interested in doing that they are of course welcome to do so. I will do what I can to help.

**Comment by adRn-s** [comment]
Thanks! Does it have an auto-update feature? Or at least a notification when there are new releases?

PS: I believe flathub/ flatpak is another great way to deliver software applications to linux in a distro-agnostic way; and it has updates. (Maybe snap too, but I prefer to distance myself from microsoft/ canonical)

**Comment by ahrm** [comment]
> Thanks! Does it have an auto-update feature? Or at least a notification when there are new releases?
> 
> PS: I believe flathub/ flatpak is another great way to deliver software applications to linux in a distro-agnostic way; and it has updates. (Maybe snap too, but I prefer to distance myself from microsoft/ canonical)

No it doesn't. I will look into flatpak.

**Comment by tinywrkb** [comment]
> > Thanks! Does it have an auto-update feature? Or at least a notification when there are new releases?
> > PS: I believe flathub/ flatpak is another great way to deliver software applications to linux in a distro-agnostic way; and it has updates. (Maybe snap too, but I prefer to distance myself from microsoft/ canonical)
> 
> No it doesn't. I will look into flatpak.

[Here's an initial PoC Flatpak packaging](https://github.com/tinywrkb/flatpaks/blob/d8b42fde3374d05d9b279464abaeb39114a0ab61/info.sioyek.sioyek/info.sioyek.sioyek.yaml).  
Too many hacks to get this packaged and running correctly.

The issues I hit while packaging this:

* I had to override `QMAKE_CC` and `QMAKE_CXX`, I believe they should default to the `gcc` and `g++` accordingly, and optionally, if exist import from the env vars `CC` and `CXX`.
* I'm choosing to build libmupdf as a shared lib, I have no idea why distro like Arch Linux package the static lib(s), this feels wrong. To get dynamic linking working and find `libmupdf` in `/app/lib`, I had to override `LIBS` and add `PKGCONFIG` var.
* Why are you installing a text file as an icon? If there's no icon, then just don't add an `Icon` key to the desktop file.  
Also, `$PREFIX/share/pixmap` is the wrong location for app icons, it shouldn't be used anymore. The correct path is `$PREFIX/share/icons/hicolor/ICON_SIZE/apps`.
* A reverse DND naming scheme should be used for the desktop file name and icon. Something like `info.sioyek.sioyek`.
* The install target of the buildsystem is partial. Package maintainers shouldn't have to manually install files (e.g. `tutorial.pdf`, default configs).
* Next to the executable is not the correct location for shaders, default configs, and docs. They should be placed somewhere like `$PREFIX/share/sioyek`.
* I had to override `MESA_GL_VERSION_OVERRIDE` during runtime to have the app render something other than a blank window. I'm not sure what's the correct solution, but this hack is not it. Related error message:
```
Compiling shader : 9��
0:1(10): error: GLSL 3.30 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.00 ES, 3.00 ES, 3.10 ES, and 3.20 ES
```  
I should add that my libpdf build might not have OpenGL support enabled.

**Comment by ahrm** [comment]
> * I'm choosing to build libmupdf as a shared lib, I have no idea why distro like Arch Linux package the static lib(s), this feels wrong. To get dynamic linking working and find `libmupdf` in `/app/lib`, I had to override `LIBS` and add `PKGCONFIG` var.

The reason that we include mupdf statically is because different mupdf versions behave differently. For example this app does not compile correctly on mupdf 1.6.* which is the one included in 18.04.

> * Why are you installing a text file as an icon? If there's no icon, then just don't add an `Icon` key to the desktop file.
>   Also, `$PREFIX/share/pixmap` is the wrong location for app icons, it shouldn't be used anymore. The correct path is `$PREFIX/share/icons/hicolor/ICON_SIZE/apps`.
> * A reverse DND naming scheme should be used for the desktop file name and icon. Something like `info.sioyek.sioyek`.
> * The install target of the buildsystem is partial. Package maintainers shouldn't have to manually install files (e.g. `tutorial.pdf`, default configs).

Yeah sorry. I am a complete newb when it comes to linux kernels so there are embarrassing issues like this.

> * Next to the executable is not the correct location for shaders, default configs, and docs. They should be placed somewhere like `$PREFIX/share/sioyek`.

I think I do put them in share/sioyek. 

> * I had to override `MESA_GL_VERSION_OVERRIDE` during runtime to have the app render something other than a blank window. I'm not sure what's the correct solution, but this hack is not it. Related error message:

Yeah another user had the same error. I think maybe it is a problem with your OpenGL driver version? I am not sure though.

**Comment by Ristovski** [comment]
> I had to override MESA_GL_VERSION_OVERRIDE during runtime to have the app render something other than a blank window

@tinywrkb Idea: check (by adding a debug print or something) the OpenGL context version. I've experienced weird errors like that before, and it was usually due to not explicitly requesting a minimum OpenGL/ES version.

**Comment by ahrm** [comment]
> > I had to override MESA_GL_VERSION_OVERRIDE during runtime to have the app render something other than a blank window
> 
> @tinywrkb Idea: check (by adding a debug print or something) the OpenGL context version. I've experienced weird errors like that before, and it was usually due to not explicitly requesting a minimum OpenGL/ES version.
> 
> Edit: It could also be due to flatpak bundling a wrong OpenGL lib or something, thus getting a wrong OpenGL version. You can also run it with `LIBGL_DEBUG=verbose ./app` and see if that tells you anything.

Thanks you. I think it should be fixed in 7025bdd71d5d623debc16e4c1752a7f073399344 then.

**Comment by tinywrkb** [comment]
> > * I'm choosing to build libmupdf as a shared lib, I have no idea why distro like Arch Linux package the static lib(s), this feels wrong. To get dynamic linking working and find `libmupdf` in `/app/lib`, I had to override `LIBS` and add `PKGCONFIG` var.
> 
> The reason that we include mupdf statically is because different mupdf versions behave differently. For example this app does not compile correctly on mupdf 1.6.* which is the one included in 18.04.

That's not a problem with the Flatpak package, as `libmupdf` is being built and packaged with the app, and it's not available in the runtime.  
Building the library with flatpak-builder as a separate module, and not in the Sioyek build tree, means that we don't have to wait on the app developer to get the latest fixes to the lib, and we can use the available automated tools to have the Flatpak package manifest updated and be built automatically.  

> > * Why are you installing a text file as an icon? If there's no icon, then just don't add an `Icon` key to the desktop file.
> >   ....
> 
> Yeah sorry. I am a complete newb when it comes to linux kernels so there are embarrassing issues like this.

No problem, this can be overwhelming as documentation and conventions are scattered.  

> > * Next to the executable is not the correct location for shaders, default configs, and docs. They should be placed somewhere like `$PREFIX/share/sioyek`.
> 
> I think I do put them in share/sioyek.

I was too lazy to looked at the code, so I just peeked in `build_linux.sh`, and IIRC the errors printed to the terminal output said that the app is looking for the missing files next to the executable in `$PREFIX/bin` (PREFIX is `/app`).

> > * I had to override `MESA_GL_VERSION_OVERRIDE` during runtime to have the app render something other than a blank window. I'm not sure what's the correct solution, but this hack is not it. Related error message:
> 
> Yeah another user had the same error. I think maybe it is a problem with your OpenGL driver version? I am not sure though.

From `glxinfo` inside the Flatpak container:

`OpenGL core profile version string: 4.6 (Core Profile) Mesa 21.1.4`

> Thanks you. I think it should be fixed in [7025bdd](https://github.com/ahrm/sioyek/commit/7025bdd71d5d623debc16e4c1752a7f073399344) then.

Yes, it's fixed now.

> > I had to override MESA_GL_VERSION_OVERRIDE during runtime to have the app render something other than a blank window
> 
> @tinywrkb Idea: check (by adding a debug print or something) the OpenGL context version. I've experienced weird errors like that before, and it was usually due to not explicitly requesting a minimum OpenGL/ES version.
> 
> Edit: It could also be due to flatpak bundling a wrong OpenGL lib or something, thus getting a wrong OpenGL version. You can also run it with `LIBGL_DEBUG=verbose ./app` and see if that tells you anything.

@Ristovski nothing really useful in the terminal output. I've been running all my desktop apps through Flatpak for a while now, even a game or two, and I haven't hit such an issue before.

**Comment by olejorgenb** [comment]
Anyone successfully build this on ubuntu. I've installed `qt5-default libharfbuzz-dev libqt5quickwidgets5 libqt53dcore5 libqt53danimation5 libqt53dextras5 libqt53drender5 libqt53dinput5 libqt53dlogic5` but the script still complains about:

> Project ERROR: Unknown module(s) in QT: quickwidgets 3dcore 3danimation 3dextras 3dinput 3dlogic 3drender

During the build of pdf_viewer..

**Comment by s1291** [comment]
@olejorgenb 
I have successfully built it on Ubuntu 20.04. You need to use the right version of make. So In my case I had to edit the script `build_linux.sh`:

```
cd mupdf
make USE_SYSTEM_HARFBUZZ=yes
cd ..
# notice the full path to qmake
/opt/Qt5.11.1/5.11.1/gcc_64/bin/qmake pdf_viewer_build_config.pro 
make

rm -r build 2> /dev/null
mkdir build
cp sioyek build/sioyek
cp pdf_viewer/prefs.config build/prefs.config
cp pdf_viewer/prefs_user.config build/prefs_user.config
cp pdf_viewer/keys.config build/keys.config
cp pdf_viewer/keys_user.config build/keys_user.config
cp -r pdf_viewer/shaders build/shaders
cp tutorial.pdf build/tutorial.pdf
```
After it finishes, run:
```
./build/sioyek
```

**Comment by szobov** [comment]
@olejorgenb  faced a similar issue, but this one helped to fix: `sudo apt install qt3d5-dev`

**Comment by CristiOlariu** [comment]
> Anyone successfully build this on ubuntu? I've installed `qt5-default libharfbuzz-dev libqt5quickwidgets5 libqt53dcore5 libqt53danimation5 libqt53dextras5 libqt53drender5 libqt53dinput5 libqt53dlogic5` but the script still complains about:
> 
> > Project ERROR: Unknown module(s) in QT: quickwidgets 3dcore 3danimation 3dextras 3dinput 3dlogic 3drender
> 
> During the build of pdf_viewer..
> 
> **EDIT:**
> 
> `qt3d5-dev` and `qtdeclarative5-private-dev` (private?! qtdeclarative5-dev not enough..) is needed. These packages contain the `qt_lib_*_private.pri` files (eg.: `qt_lib_quickwidgets_private.pri`) which apparently is needed for the `QT +=` machinery.
> 
> In case someone finds this issue having more general problems with qt modules it seems like `/usr/lib/x86_64-linux-gnu/qt5/mkspecs/modules` contain the modules available in ubuntu. (use [this](https://packages.ubuntu.com/search?searchon=contents&keywords=qt_lib_quickwidgets_private.pri&mode=exactfilename&suite=hirsute&arch=any)) to find the correct package



> @olejorgenb faced a similar issue, but this one helped to fix: `sudo apt install qt3d5-dev`

These have actually solved my problem building this on Ubuntu 20.04. Maybe 
`sudo apt install qt3d5-dev`
should be in the pre-requisites for Linux.

**Comment by nbenitez** [comment]
Sioyek it's now published on Flathub!, check out:
https://flathub.org/apps/details/com.github.ahrm.sioyek

Now it's easier to try `sioyek` in Linux no matter the distribution.
